src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -56,13 +56,30 @@ def ideal.is_homogeneous : Prop :=
 ∀ (i : ι) ⦃r : A⦄, r ∈ I → (graded_algebra.decompose 𝒜 r i : A) ∈ I
 
 /-- For any `semiring A`, we collect the homogeneous ideals of `A` into a type. -/
-abbreviation homogeneous_ideal : Type* := { I : ideal A // I.is_homogeneous 𝒜 }
+structure homogeneous_ideal extends submodule A A :=
+(is_homogeneous' : ideal.is_homogeneous 𝒜 to_submodule)
+
+variable {𝒜}
+def homogeneous_ideal.to_ideal (I : homogeneous_ideal 𝒜) : ideal A := I.to_submodule
+
+lemma homogeneous_ideal.is_homogeneous (I : homogeneous_ideal 𝒜) :
+  I.to_ideal.is_homogeneous 𝒜 := I.is_homogeneous'
+
+@[ext] lemma homogeneous_ideal.ext {I J : homogeneous_ideal 𝒜} (h : I.to_ideal = J.to_ideal) :
+  I = J := by cases I; cases J; simpa [homogeneous_ideal.to_ideal] using h
```

Comment:
This would be better stated as
```suggestion
lemma homogeneous_ideal.to_ideal_injective :
  function.injective (homogeneous_ideal.to_ideal : homogeneous_ideal 𝒜 → ideal A) :=
sorry
```